src/common/libsdprocess/sdprocess.c

```diff
@@ -0,0 +1,1835 @@
+        flux_watcher_stop (sdp->w_state_check);
+        flux_watcher_destroy (sdp->w_state_check);
+
+        free (sdp->active_state);
+        free (sdp->result);
+        free (sdp);
+        errno = save_errno;
+    }
+}
+
+static int strv_len (char **strv)
+{
+    int len = 0;
+    if (strv) {
+        char **ptr = strv;
+        while (*ptr) {
+            len++;
+            ptr++;
+        }
+    }
+    return len;
+}
+
+static int strv_copy (char **strv, char ***cpyp)
+{
+    char **cpy = NULL;
+    int len = strv_len (strv);
+
+    if (len) {
+        int i;
+        /* +1 for NULL terminating pointer */
+        if (!(cpy = calloc (1, sizeof (char *) * (len + 1))))
+            return -1;
+        for (i = 0; i < len; i++) {
+            if (!(cpy[i] = strdup (strv[i]))) {
+                strv_destroy (cpy);
+                return -1;
+            }
+        }
+        *cpyp = cpy;
+    }
+    return 0;
+}
+
+static flux_sdprocess_t *sdprocess_create (flux_t *h,
+                                           flux_reactor_t *reactor,
+                                           const char *unitname,
+                                           char **argv,
+                                           char **envv,
+                                           int stdin_fd,
+                                           int stdout_fd,
+                                           int stderr_fd)
+{
+    struct flux_sdprocess *sdp = NULL;
+    /* note that lower case should be used, not upper case hex.  Apparently
+     * a part of dbus standard.
+     */
+    char hex2char[16] = "0123456789abcdef";
+    const char *service_path_prefix = "/org/freedesktop/systemd1/unit/";
+    int service_path_prefix_len = strlen (service_path_prefix);
+    /* _2e is escape of . (period) */
+    const char *service_path_suffix = "_2eservice";
+    int service_path_suffix_len = strlen (service_path_suffix);
+    const char *ptr;
+    int index, len, ret;
+
+    if (!(sdp = calloc (1, sizeof (*sdp))))
+        return NULL;
+    if (!(sdp->unitname = strdup (unitname)))
+        goto cleanup;
+    if (strv_copy (argv, &sdp->argv) < 0)
+        goto cleanup;
+    if (strv_copy (envv, &sdp->envv) < 0)
+        goto cleanup;
+    sdp->h = h;
+    sdp->reactor = reactor;
+    sdp->stdin_fd = stdin_fd;
+    sdp->stdout_fd = stdout_fd;
+    sdp->stderr_fd = stderr_fd;
+
+    if ((ret = sd_bus_open_user (&sdp->bus)) < 0) {
+        errno = -ret;
+        goto cleanup;
+    }
+    if (asprintf (&sdp->service_name, "%s.service", sdp->unitname) < 0)
+        goto cleanup;
+    /* we must escape all special chars b/c dbus cannot handle them.
+     * escape rule is <special char> -> _<hex bytes of char>.
+     * e.g. . -> _2e
+     *
+     * note that a-f should be lower case
+     *
+     * also escape 0-9 if numeral is first character in unitname.
+     */
+    len = service_path_prefix_len;
+    len += strlen (sdp->unitname) * 3;
```

Comment:
Suggestion: break out the escaping of special characters to a "string_copy_with_escape()" function or similar.

This code is a bit hard to follow to ensure that it doesn't exceed its buffer space and that the resulting string is always null terminated.  The use of `strcpy()` tends to raise a red flag.

Reply:
(I just mean the use of `strcpy()` invites closer scrutiny - so the code that uses it should be fairly clear/commented so that the scrutiny isn't too time consuming)